### PR TITLE
[POST_LTS] avocado/utils/network.py: introduce PortTracker

### PR DESCRIPTION
**Pull request by clebergnu**
Moved from avocado-virt, this seems like a useful addition to the
avocado utility libraries.

Signed-off-by: Cleber Rosa <crosa@redhat.com>

### CODE REVIEW

**Comment by clebergnu** [comment]
Related to https://github.com/avocado-framework/avocado-virt/pull/99

File: avocado/utils/network.py [hunk 1]
```diff
@@ -71,3 +73,44 @@ def find_free_ports(start_port, end_port, count, address="localhost"):
...
+
+class PortTracker(Borg):
+
+    """
+    Tracks ports used in the host machine.
+    """
+
+    def __init__(self):
+        Borg.__init__(self)
+        self.address = 'localhost'
```
Comment (ldoktor) [inline avocado/utils/network.py]: Well this seems a bit limiting, one might be using specific address/interface and/or keep track of multiple different hosts as well (without the feature to get free ports or with a different implementation).

File: avocado/utils/network.py [hunk 2]
```diff
@@ -71,3 +73,44 @@ def find_free_ports(start_port, end_port, count, address="localhost"):
...
+        if not hasattr(self, 'retained_ports'):
+            self._reset_retained_ports()
+
+    def __str__(self):
+        return 'Ports tracked: %r' % self.retained_ports
+
+    def _reset_retained_ports(self):
+        self.retained_ports = []
+
+    def register_port(self, port):
```
Comment (ldoktor) [inline avocado/utils/network.py]: I think it'd be really useful to have `force` here to force inject it even on use (eg. when you start service, obtain address from that service and then keep track of it).

Comment (ldoktor) [inline avocado/utils/network.py]: And another idea (probably controversial), how to optionally support `note` argument, which could simplify analysis of what is occupying the port? Eg. `net.register_port(5000, "vm1's ssh redirection")`.

**Comment by lmr** [comment]
OK, I opened this tab for review before @ldoktor started his review. He makes good points, and the idea about putting notes on used ports would be a nice touch for post mortem analysis.

**Comment by clebergnu** [comment]
@ldoktor this is definitely not for immediate inclusion.  Maybe I should have made that explicit, but then, it served as a good check that we're on the same page wrt the current release goals.

With regards to the RFEs you listed, sure, the sky is the limit.  Let's keep in mind, though, that this has been working for the Avocado-Virt use case, and seems generic enough to live in Avocado instead.

**Comment by ldoktor** [comment]
Sure it worked there in quite specific usecase (to account the occupied ports while creating machines). For generic use I think some refinements would be useful. Anyway we can let it evolve inside Avocado after the LTS release.

**Comment by clebergnu** [comment]
@lmr @ldoktor you both reviewed this PR, and it looks like it's approved.  Please let me know if this needs anything else or, if not, let's merge it and get it done with.  Thanks!

**Comment by apahim** [comment]
Yes, looks good and was previously approved. Merging.

**Pull request merged**